File: src/main/python/net/coderextreme/data/force.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
meta2 = x3d.meta()
meta2.name = "creator"
meta2.content = "John W Carlson"

head1.children.append(meta2)
meta3 = x3d.meta()
meta3.name = "created"
meta3.content = "December 13 2015"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "modified"
meta4.content = "Mon, 09 Feb 2026 12:23:02 GMT"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "title"
meta5.content = "force.x3d"

head1.children.append(meta5)
meta6 = x3d.meta()
meta6.name = "identifier"
meta6.content = "https://coderextreme.net/X3DJSONLD/src/main/data/force.x3d"

head1.children.append(meta6)

X3D0.head = head1
Scene7 = x3d.Scene()
ProtoDeclare8 = x3d.ProtoDeclare()
ProtoDeclare8.name = "node"
ProtoInterface9 = x3d.ProtoInterface()
field10 = x3d.field()
field10.accessType = "inputOutput"
field10.type = "SFVec3f"
field10.name = "position"

ProtoInterface9.field.append(field10)

ProtoDeclare8.ProtoInterface = ProtoInterface9
ProtoBody11 = x3d.ProtoBody()
Group12 = x3d.Group()
Transform13 = x3d.Transform(DEF="transform")
Shape14 = x3d.Shape()
Appearance15 = x3d.Appearance()
Material16 = x3d.Material()
Material16.diffuseColor = [1,0,0]

Appearance15.material = Material16

Shape14.appearance = Appearance15
Sphere17 = x3d.Sphere()

Shape14.geometry = Sphere17

Transform13.children.append(Shape14)
Transform18 = x3d.Transform()
Transform18.translation = [1,0,0]
Shape19 = x3d.Shape()
Appearance20 = x3d.Appearance()
Material21 = x3d.Material()
Material21.diffuseColor = [0,0,1]

Appearance20.material = Material21

Shape19.appearance = Appearance20
Text22 = x3d.Text()
Text22.string = ["Node"]
FontStyle23 = x3d.FontStyle()
FontStyle23.size = 5
FontStyle23.justify = ["MIDDLE","MIDDLE"]

Text22.fontStyle = FontStyle23

Shape19.geometry = Text22

Transform18.children.append(Shape19)

Transform13.children.append(Transform18)
IS24 = x3d.IS()
connect25 = x3d.connect()
connect25.nodeField = "translation"
connect25.protoField = "position"

IS24.connect.append(connect25)

Transform13.IS = IS24

Group12.children.append(Transform13)
PositionInterpolator26 = x3d.PositionInterpolator(DEF="NodePosition")
PositionInterpolator26.key = [0,1]
PositionInterpolator26.keyValue = [(0, 0, 0),(0, 5, 0)]

Group12.children.append(PositionInterpolator26)
Script27 = x3d.Script(DEF="MoveBall")
field28 = x3d.field()
field28.accessType = "inputOutput"
field28.type = "SFVec3f"
field28.name = "translation"
field28.value = [50,50,0]

Script27.field.append(field28)
field29 = x3d.field()
field29.accessType = "inputOutput"
field29.type = "SFVec3f"
field29.name = "old"

Script27.field.append(field29)
field30 = x3d.field()
field30.accessType = "inputOnly"
field30.type = "SFTime"
field30.name = "set_cycle"

Script27.field.append(field30)
field31 = x3d.field()
field31.accessType = "outputOnly"
field31.type = "MFVec3f"
field31.name = "keyValue"

Script27.field.append(field31)

Script27.sourceCode = '''ecmascript:\n"+
"			function set_cycle(value) {\n"+
"				old = translation;\n"+
"				translation = new SFVec3f(Math.random()*100-50, Math.random()*100-50, Math.random()*100-50);\n"+
"				keyValue = new MFVec3f(old, translation);\n"+
"			}'''

Group12.children.append(Script27)
TimeSensor32 = x3d.TimeSensor(DEF="nodeClock")
TimeSensor32.cycleInterval = 3
TimeSensor32.loop = True

Group12.children.append(TimeSensor32)

ProtoBody11.children.append(Group12)
ROUTE33 = x3d.ROUTE()
ROUTE33.fromNode = "nodeClock"
ROUTE33.fromField = "cycleTime"
ROUTE33.toNode = "MoveBall"
ROUTE33.toField = "set_cycle"

ProtoBody11.children.append(ROUTE33)
ROUTE34 = x3d.ROUTE()
ROUTE34.fromNode = "nodeClock"
ROUTE34.fromField = "fraction_changed"
ROUTE34.toNode = "NodePosition"
ROUTE34.toField = "set_fraction"

ProtoBody11.children.append(ROUTE34)
ROUTE35 = x3d.ROUTE()
ROUTE35.fromNode = "MoveBall"
ROUTE35.fromField = "keyValue"
ROUTE35.toNode = "NodePosition"
ROUTE35.toField = "set_keyValue"

ProtoBody11.children.append(ROUTE35)
ROUTE36 = x3d.ROUTE()
ROUTE36.fromNode = "NodePosition"
ROUTE36.fromField = "value_changed"
ROUTE36.toNode = "transform"
ROUTE36.toField = "set_translation"

ProtoBody11.children.append(ROUTE36)

ProtoDeclare8.ProtoBody = ProtoBody11

Scene7.children.append(ProtoDeclare8)
ProtoDeclare37 = x3d.ProtoDeclare()
ProtoDeclare37.name = "cyl"
ProtoInterface38 = x3d.ProtoInterface()
field39 = x3d.field()
field39.accessType = "inputOutput"
field39.type = "SFVec3f"
field39.name = "set_positionA"

ProtoInterface38.field.append(field39)
field40 = x3d.field()
field40.accessType = "inputOutput"
field40.type = "SFVec3f"
field40.name = "set_positionB"
field40.value = [50,50,50]

ProtoInterface38.field.append(field40)

ProtoDeclare37.ProtoInterface = ProtoInterface38
ProtoBody41 = x3d.ProtoBody()
Group42 = x3d.Group()
Shape43 = x3d.Shape()
Appearance44 = x3d.Appearance()
Material45 = x3d.Material()
Material45.diffuseColor = [0,1,0]

Appearance44.material = Material45

Shape43.appearance = Appearance44
Extrusion46 = x3d.Extrusion(DEF="extrusion")
Extrusion46.creaseAngle = 0.785
Extrusion46.crossSection = [(1, 0),(0.92, -0.38),(0.71, -0.71),(0.38, -0.92),(0, -1),(-0.38, -0.92),(-0.71, -0.71),(-0.92, -0.38),(-1, 0),(-0.92, 0.38),(-0.71, 0.71),(-0.38, 0.92),(0, 1),(0.38, 0.92),(0.71, 0.71),(0.92, 0.38),(1, 0)]
Extrusion46.spine = [(0, -50, 0),(0, 0, 0),(0, 50, 0)]

Shape43.geometry = Extrusion46

Group42.children.append(Shape43)
Script47 = x3d.Script(DEF="MoveCylinder")
field48 = x3d.field()
field48.accessType = "inputOutput"
field48.type = "MFVec3f"
field48.name = "spine"
field48.value = [(0, -50, 0),(0, 0, 0),(0, 50, 0)]

Script47.field.append(field48)
field49 = x3d.field()
field49.accessType = "inputOutput"
field49.type = "SFVec3f"
field49.name = "endA"

Script47.field.append(field49)
field50 = x3d.field()
field50.accessType = "inputOutput"
field50.type = "SFVec3f"
field50.name = "endB"

Script47.field.append(field50)

Script47.sourceCode = '''ecmascript:\n"+
"			function set_endA(value) {\n"+
"				spine = new MFVec3f(value, spine[1]);\n"+
"			}\n"+
"\n"+
"			function set_endB(value) {\n"+
"				spine = new MFVec3f(spine[0], value);\n"+
"			}'''
IS51 = x3d.IS()
connect52 = x3d.connect()
connect52.nodeField = "endA"
connect52.protoField = "set_positionA"

IS51.connect.append(connect52)
connect53 = x3d.connect()
connect53.nodeField = "endB"
connect53.protoField = "set_positionB"

IS51.connect.append(connect53)

Script47.IS = IS51

Group42.children.append(Script47)

ProtoBody41.children.append(Group42)
ROUTE54 = x3d.ROUTE()
ROUTE54.fromNode = "MoveCylinder"
ROUTE54.fromField = "spine_changed"
ROUTE54.toNode = "extrusion"
ROUTE54.toField = "set_spine"

ProtoBody41.children.append(ROUTE54)

ProtoDeclare37.ProtoBody = ProtoBody41

Scene7.children.append(ProtoDeclare37)
Transform55 = x3d.Transform(DEF="HoldsContent")
Transform55.scale = [0.1,0.1,0.1]
PlaneSensor56 = x3d.PlaneSensor(DEF="clickGenerator")
PlaneSensor56.description = "click on background to add nodes, click on nodes to add links"
PlaneSensor56.minPosition = [-50,-50]
PlaneSensor56.maxPosition = [50,50]

Transform55.children.append(PlaneSensor56)
ProtoInstance57 = x3d.ProtoInstance(DEF="nodeA")
ProtoInstance57.name = "node"
fieldValue58 = x3d.fieldValue()
fieldValue58.name = "position"
fieldValue58.value = "0 4.483333 0"

ProtoInstance57.fieldValue.append(fieldValue58)

Transform55.children.append(ProtoInstance57)
ProtoInstance59 = x3d.ProtoInstance(DEF="nodeB")
ProtoInstance59.name = "node"
fieldValue60 = x3d.fieldValue()
fieldValue60.name = "position"
fieldValue60.value = "0 4.483333 0"

ProtoInstance59.fieldValue.append(fieldValue60)

Transform55.children.append(ProtoInstance59)
ProtoInstance61 = x3d.ProtoInstance(DEF="nodeC")
ProtoInstance61.name = "node"
fieldValue62 = x3d.fieldValue()
fieldValue62.name = "position"
fieldValue62.value = "0 4.483333 0"

ProtoInstance61.fieldValue.append(fieldValue62)

Transform55.children.append(ProtoInstance61)
ProtoInstance63 = x3d.ProtoInstance(DEF="nodeD")
ProtoInstance63.name = "node"
fieldValue64 = x3d.fieldValue()
fieldValue64.name = "position"
fieldValue64.value = "0 4.483333 0"

ProtoInstance63.fieldValue.append(fieldValue64)

Transform55.children.append(ProtoInstance63)
ProtoInstance65 = x3d.ProtoInstance(DEF="linkA")
ProtoInstance65.name = "cyl"
fieldValue66 = x3d.fieldValue()
fieldValue66.name = "set_positionA"
fieldValue66.value = "0 4.483333 0"

ProtoInstance65.fieldValue.append(fieldValue66)
fieldValue67 = x3d.fieldValue()
fieldValue67.name = "set_positionB"
fieldValue67.value = "0 4.483333 0"

ProtoInstance65.fieldValue.append(fieldValue67)

Transform55.children.append(ProtoInstance65)
ProtoInstance68 = x3d.ProtoInstance(DEF="linkB")
ProtoInstance68.name = "cyl"
fieldValue69 = x3d.fieldValue()
fieldValue69.name = "set_positionA"
fieldValue69.value = "0 4.483333 0"

ProtoInstance68.fieldValue.append(fieldValue69)
fieldValue70 = x3d.fieldValue()
fieldValue70.name = "set_positionB"
fieldValue70.value = "0 4.483333 0"

ProtoInstance68.fieldValue.append(fieldValue70)

Transform55.children.append(ProtoInstance68)
ProtoInstance71 = x3d.ProtoInstance(DEF="linkC")
ProtoInstance71.name = "cyl"
fieldValue72 = x3d.fieldValue()
fieldValue72.name = "set_positionA"
fieldValue72.value = "0 4.483333 0"

ProtoInstance71.fieldValue.append(fieldValue72)
fieldValue73 = x3d.fieldValue()
fieldValue73.name = "set_positionB"
fieldValue73.value = "0 4.483333 0"

ProtoInstance71.fieldValue.append(fieldValue73)

Transform55.children.append(ProtoInstance71)

Scene7.children.append(Transform55)
Script74 = x3d.Script(DEF="clickHandler")
field75 = x3d.field()
field75.accessType = "inputOutput"
field75.type = "SFInt32"
field75.name = "counter"

Script74.field.append(field75)
field76 = x3d.field()
field76.accessType = "inputOnly"
field76.type = "SFBool"
field76.name = "add_node"

Script74.field.append(field76)

Script74.sourceCode = '''ecmascript:\n"+
"	function add_node(value) {\n"+
"                // Browser.print('hey ', counter);\n"+
"                counter = counter++;\n"+
"		Browser.appendTo(Browser.getDocument().querySelector(\"field [name=ModifiableNode]\"),\n"+
"			{ \"ProtoInstance\":\n"+
"				{ \"@name\":\"node\",\n"+
"				  \"@DEF\":\"node'+counter+'\",\n"+
"				  \"fieldValue\": [\n"+
"					{\n"+
"						 \"@name\":\"position\",\n"+
"						 \"@value\":[0.0,0.0,0.0]\n"+
"					}\n"+
"				  ]\n"+
"				}\n"+
"			});\n"+
"                \n"+
"        }'''

Scene7.children.append(Script74)
ROUTE77 = x3d.ROUTE()
ROUTE77.fromNode = "clickGenerator"
ROUTE77.fromField = "isActive"
ROUTE77.toNode = "clickHandler"
ROUTE77.toField = "add_node"

Scene7.children.append(ROUTE77)
ROUTE78 = x3d.ROUTE()
ROUTE78.fromNode = "nodeA"
ROUTE78.fromField = "position_changed"
ROUTE78.toNode = "linkA"
ROUTE78.toField = "set_set_positionA"

Scene7.children.append(ROUTE78)
ROUTE79 = x3d.ROUTE()
ROUTE79.fromNode = "nodeB"
ROUTE79.fromField = "position_changed"
ROUTE79.toNode = "linkA"
ROUTE79.toField = "set_set_positionB"

Scene7.children.append(ROUTE79)
ROUTE80 = x3d.ROUTE()
ROUTE80.fromNode = "nodeA"
ROUTE80.fromField = "position_changed"
ROUTE80.toNode = "linkB"
ROUTE80.toField = "set_set_positionA"

Scene7.children.append(ROUTE80)
ROUTE81 = x3d.ROUTE()
ROUTE81.fromNode = "nodeC"
ROUTE81.fromField = "position_changed"
ROUTE81.toNode = "linkB"
ROUTE81.toField = "set_set_positionB"

Scene7.children.append(ROUTE81)
ROUTE82 = x3d.ROUTE()
ROUTE82.fromNode = "nodeA"
ROUTE82.fromField = "position_changed"
ROUTE82.toNode = "linkC"
ROUTE82.toField = "set_set_positionA"

Scene7.children.append(ROUTE82)
ROUTE83 = x3d.ROUTE()
ROUTE83.fromNode = "nodeD"
ROUTE83.fromField = "position_changed"
ROUTE83.toNode = "linkC"
ROUTE83.toField = "set_set_positionB"

Scene7.children.append(ROUTE83)

X3D0.Scene = Scene7
f = open("../data/force.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/force.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/force.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/dart/net/coderextreme/data/force.dart
// Copyright 2021 by John Carlson.  All rights reserved.
import 'dart:io';
import '../x3d.dart';
var X3D0 =  X3D(

      profile_ : SFString('Immersive'),
      version_ : SFString('3.3'),
      head_ : 
        head(
          component_ : 
            component(
              name_ : SFString('Scripting'),
              level_ : 1),
          meta_ : [
            meta(
              name_ : SFString('creator'),
              content_ : SFString('John W Carlson')),

            meta(
              name_ : SFString('created'),
              content_ : SFString('December 13 2015')),

            meta(
              name_ : SFString('title'),
              content_ : SFString('force.x3d')),

            meta(
              name_ : SFString('identifier'),
              content_ : SFString('https://coderextreme.net/X3DJSONLD/src/main/data/force.x3d')),

            meta(
              name_ : SFString('description'),
              content_ : SFString('beginnings of a force directed graph in 3D')),

            meta(
              name_ : SFString('generator'),
              content_ : SFString('Vim, X3D-Edit, https://savage.nps.edu/X3D-Edit'))]),
      Scene_ : 
        Scene(
          children_ : [
            ProtoDeclare(
              name_ : SFString('node'),
              ProtoInterface_ : 
                ProtoInterface(
                  field_ : [
                    field(
                      type_ : SFString("SFVec3f"),
                      name_ : SFString('position'),
                      accessType_ : SFString("inputOutput"),
                      value_ : SFString('0 0 0'))]),
              ProtoBody_ : 
                ProtoBody(
                  children_ : [
                    Group(
                      children_ : [
                        Transform(
                          DEF_ : SFString('transform'),
                          IS_ : 
                            IS(
                              connect_ : [
                                connect(
                                  nodeField_ : SFString('translation'),
                                  protoField_ : SFString('position'))]),
                          child_ : 
                            Shape(
                              geometry_ : 
                                Sphere(),
                              appearance_ : 
                                Appearance(
                                  material_ : 
                                    Material(
                                      diffuseColor_ : SFColor([SFDouble(1), SFDouble(0), SFDouble(0)])))),
                          children_ : [
                            Transform(
                              translation_ : SFVec3f([SFDouble(1), SFDouble(0), SFDouble(0)]),
                              child_ : 
                                Shape(
                                  geometry_ : 
                                    Text(
                                      string_ : MFString([SFString("Node")]),
                                      fontStyle_ : 
                                        FontStyle(
                                          justify_ : MFString([SFString("MIDDLE"), SFString("MIDDLE")]),
                                          size_ : 5)),
                                  appearance_ : 
                                    Appearance(
                                      material_ : 
                                        Material(
                                          diffuseColor_ : SFColor([SFDouble(0), SFDouble(0), SFDouble(1)])))))]),

                        PositionInterpolator(
                          DEF_ : SFString('NodePosition'),
                          key_ : MFFloat([SFFloat(0), SFFloat(1)]),
                          keyValue_ : MFVec3f([SFVec3f([0,0,0]),SFVec3f([0,5,0])])),

                        Script(
                          DEF_ : SFString('MoveBall'),
                          field_ : [
                            field(
                              type_ : SFString("SFVec3f"),
                              name_ : SFString('translation'),
                              accessType_ : SFString("inputOutput"),
                              value_ : SFString('50 50 0')),

                            field(
                              type_ : SFString("SFVec3f"),
                              name_ : SFString('old'),
                              accessType_ : SFString("inputOutput"),
                              value_ : SFString('0 0 0')),

                            field(
                              type_ : SFString("SFTime"),
                              name_ : SFString('set_cycle'),
                              accessType_ : SFString("inputOnly")),

                            field(
                              type_ : SFString("MFVec3f"),
                              name_ : SFString('keyValue'),
                              accessType_ : SFString("outputOnly")),
                          ],
ecmascript:eval (0
					, function set_cycle(value) {
                                                old = translation;
						translation = new SFVec3f(Math.random()*100-50, Math.random()*100-50, Math.random()*100-50);
                                                keyValue = new MFVec3f([old, translation]);
						// Browser.println(translation);
					})),

                        TimeSensor(
                          DEF_ : SFString('nodeClock'),
                          cycleInterval_ : 3,
                          loop_ : true),

                        ROUTE(
                          fromNode_ : SFString('nodeClock'),
                          fromField_ : SFString('cycleTime'),
                          toNode_ : SFString('MoveBall'),
                          toField_ : SFString('set_cycle')),

                        ROUTE(
                          fromNode_ : SFString('nodeClock'),
                          fromField_ : SFString('fraction_changed'),
                          toNode_ : SFString('NodePosition'),
                          toField_ : SFString('set_fraction')),

                        ROUTE(
                          fromNode_ : SFString('MoveBall'),
                          fromField_ : SFString('keyValue'),
                          toNode_ : SFString('NodePosition'),
                          toField_ : SFString('keyValue')),

                        ROUTE(
                          fromNode_ : SFString('NodePosition'),
                          fromField_ : SFString('value_changed'),
                          toNode_ : SFString('transform'),
                          toField_ : SFString('set_translation'))])])),

            ProtoDeclare(
              name_ : SFString('cylinder'),
              ProtoInterface_ : 
                ProtoInterface(
                  field_ : [
                    field(
                      type_ : SFString("SFVec3f"),
                      name_ : SFString('set_positionA'),
                      accessType_ : SFString("inputOnly")),

                    field(
                      type_ : SFString("SFVec3f"),
                      name_ : SFString('set_positionB'),
                      accessType_ : SFString("inputOnly"))]),
              ProtoBody_ : 
                ProtoBody(
                  children_ : [
                    Group(
                      children_ : [
                        Shape(
                          geometry_ : 
                            Extrusion(
                              DEF_ : SFString('extrusion'),
                              creaseAngle_ : 0.785,
                              crossSection_ : MFVec2f([SFVec2f([1,0]),SFVec2f([0.92,-0.38]),SFVec2f([0.71,-0.71]),SFVec2f([0.38,-0.92]),SFVec2f([0,-1]),SFVec2f([-0.38,-0.92]),SFVec2f([-0.71,-0.71]),SFVec2f([-0.92,-0.38]),SFVec2f([-1,0]),SFVec2f([-0.92,0.38]),SFVec2f([-0.71,0.71]),SFVec2f([-0.38,0.92]),SFVec2f([0,1]),SFVec2f([0.38,0.92]),SFVec2f([0.71,0.71]),SFVec2f([0.92,0.38]),SFVec2f([1,0])]),
                              spine_ : MFVec3f([SFVec3f([0,-50,0]),SFVec3f([0,50,0])])),
                          appearance_ : 
                            Appearance(
                              material_ : 
                                Material(
                                  diffuseColor_ : SFColor([SFDouble(0), SFDouble(1), SFDouble(0)])))),

                        Script(
                          DEF_ : SFString('MoveCylinder'),
                          field_ : [
                            field(
                              type_ : SFString("MFVec3f"),
                              name_ : SFString('spine'),
                              accessType_ : SFString("inputOutput"),
                              value_ : SFString('0 -50 0 0 50 0')),

                            field(
                              type_ : SFString("SFVec3f"),
                              name_ : SFString('set_endA'),
                              accessType_ : SFString("inputOnly")),

                            field(
                              type_ : SFString("SFVec3f"),
                              name_ : SFString('set_endB'),
                              accessType_ : SFString("inputOnly"))],

                          IS_ : 
                            IS(
                              connect_ : [
                                connect(
                                  nodeField_ : SFString('set_endA'),
                                  protoField_ : SFString('set_positionA')),

                                connect(
                                  nodeField_ : SFString('set_endB'),
                                  protoField_ : SFString('set_positionB'))]),
                          ,
ecmascript:eval (0

                , function set_endA(value) {
		    if (typeof spine === 'undefined') {
		        spine = new MFVec3f([value, value]);
		    } else {
		        spine = new MFVec3f([value, spine[1]]);
		    }
                }

                , function set_endB(value) {
		    if (typeof spine === 'undefined') {
		        spine = new MFVec3f([value, value]);
		    } else {
		        spine = new MFVec3f([spine[0], value]);
		    }
                }

                , function set_spine(value) {
                    spine = value;
                })),

                        ROUTE(
                          fromNode_ : SFString('MoveCylinder'),
                          fromField_ : SFString('spine'),
                          toNode_ : SFString('extrusion'),
                          toField_ : SFString('set_spine'))])])),

            Transform(
              DEF_ : SFString('HoldsContent'),
              scale_ : SFVec3f([SFDouble(0.1), SFDouble(0.1), SFDouble(0.1)]),
              children_ : [
                PlaneSensor(
                  DEF_ : SFString('clickGenerator'),
                  minPosition_ : SFVec2f([SFDouble(-50), SFDouble(-50)]),
                  maxPosition_ : SFVec2f([SFDouble(50), SFDouble(50)]),
                  description_ : SFString('click on background to add nodes, click on nodes to add links')),

                ProtoInstance(
                  name_ : SFString('node'),
                  DEF_ : SFString('nodeA'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('position'),
                      value_ : SFString('0 0 0'))]),

                ProtoInstance(
                  name_ : SFString('node'),
                  DEF_ : SFString('nodeB'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('position'),
                      value_ : SFString('50 50 50'))]),

                ProtoInstance(
                  name_ : SFString('node'),
                  DEF_ : SFString('nodeC'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('position'),
                      value_ : SFString('-50 -50 -50'))]),

                ProtoInstance(
                  name_ : SFString('node'),
                  DEF_ : SFString('nodeD'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('position'),
                      value_ : SFString('50 50 -50'))]),

                ProtoInstance(
                  name_ : SFString('cylinder'),
                  DEF_ : SFString('linkA'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('set_positionA'),
                      value_ : SFString('0 0 0')),

                    fieldValue(
                      name_ : SFString('set_positionB'),
                      value_ : SFString('50 50 50'))]),

                ProtoInstance(
                  name_ : SFString('cylinder'),
                  DEF_ : SFString('linkB'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('set_positionA'),
                      value_ : SFString('0 0 0')),

                    fieldValue(
                      name_ : SFString('set_positionB'),
                      value_ : SFString('-50 -50 -50'))]),

                ProtoInstance(
                  name_ : SFString('cylinder'),
                  DEF_ : SFString('linkC'),
                  fieldValue_ : [
                    fieldValue(
                      name_ : SFString('set_positionA'),
                      value_ : SFString('50 50 50')),

                    fieldValue(
                      name_ : SFString('set_positionB'),
                      value_ : SFString('50 50 -50'))])]),

            Script(
              DEF_ : SFString('clickHandler'),
              field_ : [
                field(
                  type_ : SFString("SFInt32"),
                  name_ : SFString('counter'),
                  accessType_ : SFString("inputOutput"),
                  value_ : SFString('0')),

                field(
                  type_ : SFString("SFNode"),
                  name_ : SFString('node_changed'),
                  accessType_ : SFString("outputOnly")),

                field(
                  type_ : SFString("SFBool"),
                  name_ : SFString('add_node'),
                  accessType_ : SFString("inputOnly"),
                  value_ : SFString('false')),
              /*<field name=\"ModifiableNode\" type=\"SFNode\" accessType=\"inputOutput\"> <Transform USE=\"HoldsContent\"/> </field>*/
              ],
ecmascript:eval (0
	, function add_node(value) {
                // Browser.print('hey ', counter);
                counter = counter++;
		Browser.appendTo(Browser.getDocument().querySelector("field [name=ModifiableNode]"),
			{ "ProtoInstance":
				{ "@name":"node",
				  "@DEF":"node'+counter+'",
				  "fieldValue": [
					{
						 "@name":"position",
						 "@value":[0.0,0.0,0.0]
					}
				  ]
				}
			});

        })),

            ROUTE(
              fromNode_ : SFString('clickGenerator'),
              fromField_ : SFString('isActive'),
              toNode_ : SFString('clickHandler'),
              toField_ : SFString('add_node')),

            ROUTE(
              fromNode_ : SFString('nodeA'),
              fromField_ : SFString('position'),
              toNode_ : SFString('linkA'),
              toField_ : SFString('set_positionA')),

            ROUTE(
              fromNode_ : SFString('nodeB'),
              fromField_ : SFString('position'),
              toNode_ : SFString('linkA'),
              toField_ : SFString('set_positionB')),

            ROUTE(
              fromNode_ : SFString('nodeA'),
              fromField_ : SFString('position'),
              toNode_ : SFString('linkB'),
              toField_ : SFString('set_positionA')),

            ROUTE(
              fromNode_ : SFString('nodeC'),
              fromField_ : SFString('position'),
              toNode_ : SFString('linkB'),
              toField_ : SFString('set_positionB')),

            ROUTE(
              fromNode_ : SFString('nodeA'),
              fromField_ : SFString('position'),
              toNode_ : SFString('linkC'),
              toField_ : SFString('set_positionA')),

            ROUTE(
              fromNode_ : SFString('nodeD'),
              fromField_ : SFString('position'),
              toNode_ : SFString('linkC'),
              toField_ : SFString('set_positionB'))]));
void main() { exit(0); }
